Extractor/Connect/SessionManager.cs

```diff
@@ -0,0 +1,556 @@
+                    {
+                        triggerReconnect = false;
+                        log.LogDebug(
+                            "Last switch was too recent ({Last}), waiting until the next trigger after {Next}",
+                            lastConnectionSwitch.Value, lastConnectionSwitch + config.Redundancy.ReconnectIntervalValue.Value);
+                    }
+                }
+
+                if (oldServiceLevel >= config.Redundancy.ServiceLevelThreshold)
+                {
+                    log.LogWarning("Service level dropped below threshold. Until it is recovered, the extractor will not update history state");
+                    await client.Callbacks.OnServicelevelBelowThreshold(client);
+                }
+            }
+            else if (oldServiceLevel < config.Redundancy.ServiceLevelThreshold)
+            {
+                log.LogInformation("New service level {Level} is a above threshold {Threshold}, triggering callback", newLevel, config.Redundancy.ServiceLevelThreshold);
+                await client.Callbacks.OnServiceLevelAboveThreshold(client);
+            }
+
+            CurrentServiceLevel = newLevel;
+            ConnectionUtils.ServiceLevel.Set(CurrentServiceLevel);
+
+            return triggerReconnect;
+        }
+
+        private async Task<bool> HandleConnectResult(ConnectResult result, CancellationToken token)
+        {
+            // No reconnect, do nothing here.
+            if (result.Type == ConnectType.None) return false;
+            ConnectionUtils.Connected.Set(1);
+
+            // Ensure that we're correctly monitoring service level.
+            if (config.Redundancy.MonitorServiceLevel)
+            {
+                EnsureServiceLevelSubscription();
+                if (config.Redundancy.ReconnectIntervalValue.Value != Timeout.InfiniteTimeSpan
+                    && !client.Callbacks.TaskScheduler.ContainsTask("CheckServiceLevel"))
+                {
+                    client.Callbacks.TaskScheduler.SchedulePeriodicTask("CheckServiceLevel",
+                        config.Redundancy.ReconnectIntervalValue,
+                        async (token) =>
+                        {
+                            using (await stateLock.LockAsync(token))
+                            {
+                                pendingActions.RecheckServiceLevel = true;
+                                onNewPendingActions.NotifyAll();
+                            }
+                        }
+                    );
+                }
+            }
+
+            // Fetch the current service level as well, and trigger an update.
+            if (config.IsRedundancyEnabled || config.Redundancy.MonitorServiceLevel)
+            {
+                var sl = await ReadServiceLevel(result.Connection.Session, token);
+                var slResult = UpdateServiceLevel(sl, true);
+            }
+
+            if (connection?.Session != result.Connection.Session)
```

Comment:
Would it make sense to compare by the SessionId property?

Reply:
Not for this code. This is important for guarding against memory leaks, here I actually do care about whether we have a new _Session_ object, not whether we have a new OPC-UA session.